presto-main/src/main/java/com/facebook/presto/util/LowMemoryMonitor.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.util;
+
+import com.facebook.airlift.log.Logger;
+import com.facebook.presto.execution.TaskManagerConfig;
+import com.facebook.presto.execution.executor.TaskExecutor;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.MemoryMXBean;
+import java.lang.management.MemoryUsage;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static com.facebook.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newScheduledThreadPool;
+
+public class LowMemoryMonitor
+{
+    private static final Logger log = Logger.get(LowMemoryMonitor.class);
+    private final ScheduledExecutorService lowMemoryExecutor = newScheduledThreadPool(1, daemonThreadsNamed("worker-low-memory-executor"));
+
+    private final TaskExecutor taskExecutor;
+    private final double threshold;
+
+    @Inject
+    public LowMemoryMonitor(TaskExecutor taskExecutor, TaskManagerConfig taskManagerConfig)
+    {
+        this.taskExecutor = requireNonNull(taskExecutor, "taskExecutor is null");
+        this.threshold = taskManagerConfig.getMemoryBasedSlowDownThreshold();
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (threshold < 1.0) {
+            lowMemoryExecutor.scheduleWithFixedDelay(() -> checkLowMemory(), 1, 1, TimeUnit.SECONDS);
+        }
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        lowMemoryExecutor.shutdown();
+    }
+
+    public void checkLowMemory()
+    {
+        MemoryMXBean mbean = ManagementFactory.getMemoryMXBean();
+        MemoryUsage memoryUsage = mbean.getHeapMemoryUsage();
+
+        long usedMemory = memoryUsage.getUsed();
+        long maxMemory = memoryUsage.getMax();
+        long memoryThreshold = (long) (maxMemory * threshold);
+
+        if (usedMemory > memoryThreshold) {
+            if (!taskExecutor.isLowMemory()) {
```

Comment:
It maybe simpler to have the TaskExecutor do its own state keeping for low memory mode. It can ensure the idempotency you are shooting for in addition to the logging. So the call can become simpler here in terms of taskExecutor.setLowMemory(usedMemory, maxMemory, memoryThresh). 

It should be a low overhead call since anyway it is done once every K milliseconds